src/dsa/rpo_falcon512/error.rs

```diff
@@ -0,0 +1,51 @@
+use super::{LOG_N, MODULUS, PK_LEN};
+use core::fmt;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum FalconError {
+    KeyGenerationFailed,
+    PubKeyDecodingExtraData,
+    PubKeyDecodingInvalidCoefficient(u32),
+    PubKeyDecodingInvalidLength(usize),
+    PubKeyDecodingInvalidTag(u8),
+    SigDecodingTooBigHighBits(u32),
+    SigDecodingInvalidRemainder,
+    SigDecodingNonZeroUnusedBitsLastByte,
+    SigDecodingMinusZero,
+    SigGenerationFailed,
+}
+
+impl fmt::Display for FalconError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use FalconError::*;
+        match self {
+            KeyGenerationFailed => write!(f, "Failed to generate a private-public key pair"),
+            PubKeyDecodingExtraData => {
+                write!(f, "Failed to decode public key: input not fully consumed")
+            }
+            PubKeyDecodingInvalidCoefficient(val) => {
+                write!(f, "Failed to decode public key: coefficient {val} is greater than or equal to the field modulus {MODULUS}")
+            }
+            PubKeyDecodingInvalidLength(len) => {
+                write!(f, "Failed to decode public key: expected {PK_LEN} bytes but received {len}")
+            }
+            PubKeyDecodingInvalidTag(byte) => {
+                write!(f, "Failed to decode public key: expected the first byte to be {LOG_N} but was {byte}")
+            }
+            SigDecodingTooBigHighBits(m) => {
+                write!(f, "Failed to decode signature: high bits {m} exceed 2048")
+            }
+            SigDecodingInvalidRemainder => {
+                write!(f, "Failed to decode signature: incorrect remaining data")
+            }
+            SigGenerationFailed => write!(f, "Failed to generate a signature"),
+            SigDecodingNonZeroUnusedBitsLastByte => {
+                write!(f, "Non-zero unused bits in the last byte")
+            }
```

Comment:
nit: I'd move this up so that the branches are sorted alphabetically. Also, I'd prefix the message with "Failed to decode signature:".